test/test-utils/src/main/java/software/amazon/awssdk/testutils/Waiter.java

```diff
@@ -0,0 +1,131 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.testutils;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletionException;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * This retries a particular function multiple times until it returns an expected result (or fails with an exception). Certain
+ * expected exception types can be ignored.
+ */
+public final class Waiter<T> {
+    private static final Logger log = Logger.loggerFor(Waiter.class);
+
+    private final Supplier<T> thingToTry;
+    private Predicate<T> whenToStop = t -> true;
+    private Set<Class<? extends Throwable>> whatToIgnore = Collections.emptySet();
+
+    /**
+     * @see #run(Supplier)
+     */
+    private Waiter(Supplier<T> thingToTry) {
+        Validate.paramNotNull(thingToTry, "thingToTry");
+        this.thingToTry = thingToTry;
+    }
+
+    /**
+     * Create a waiter that attempts executing the provided function until the condition set with {@link #until(Predicate)} is
+     * met or until it throws an exception. Expected exception types can be ignored with {@link #ignoring(Class[])}.
+     */
+    public static <T> Waiter<T> run(Supplier<T> thingToTry) {
+        return new Waiter<>(thingToTry);
+    }
+
+    /**
+     * Define the condition under which the thing we are trying is complete.
+     *
+     * If this isn't set, it will always be true. ie. if the function call succeeds, we stop waiting.
+     */
+    public Waiter<T> until(Predicate<T> whenToStop) {
+        this.whenToStop = whenToStop;
+        return this;
+    }
+
+    /**
+     * Define the exception types that should be ignored if the thing we are trying throws them.
+     */
+    @SafeVarargs
+    public final Waiter<T> ignoring(Class<? extends Throwable>... whatToIgnore) {
+        this.whatToIgnore = new HashSet<>(Arrays.asList(whatToIgnore));
+        return this;
+    }
+
+    /**
+     * Execute the function, throwing an assertion error if the thing we're trying does not succeed after 30 seconds.
+     */
+    public T orFail() {
+        return orFailAfter(Duration.ofSeconds(30));
+    }
+
+    /**
+     * Execute the function, throwing an assertion error if the thing we're trying does not succeed after the provided duration.
+     */
+    public T orFailAfter(Duration howLongToTry) {
+        Validate.paramNotNull(howLongToTry, "howLongToTry");
+
+        Instant start = Instant.now();
+        int attempt = 0;
+
+        while (Duration.between(start, Instant.now()).compareTo(howLongToTry) < 0) {
+            ++attempt;
+            try {
+                if (attempt > 0) {
```

Comment:
Won't this always be true? You init attempt to 0, then add 1 to it on line 93 - so even on the very first run it'll be > 0...

Reply:
Good catch.

Reply:
which means we're always waiting 2000 ms before we even attempt it right?